### PR TITLE
Feat/louis/arb for mgv v2

### PR DESCRIPTION
**Pull request by Louis-Amas**

### CODE REVIEW

File: packages/bot-arbitrage/src/arbitarger.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      return token.allowance({
+        owner: arbitragerContract.address,
+        spender: mgv.address,
+      });
+    })
+  );
+
+  const tokenToActivate = values
+    .map((value, index) => ({ value, index }))
+    .filter((value) => value.value.eq(0))
```
Comment (maxencerb) [inline packages/bot-arbitrage/src/arbitarger.ts:21]: We assume that not 0 is infinite approval ?

File: packages/bot-arbitrage/src/arbitarger.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  const gasEstimate = await arbitragerContract.estimateGas.activateTokens(
+    tokenToActivate.map((token) => token.address)
+  );
+
+  const tx = await arbitragerContract.activateTokens(
+    tokenToActivate.map((token) => token.address),
+    {
+      gasLimit: gasEstimate.mul(10),
+    }
+  );
```
Comment (maxencerb) [inline packages/bot-arbitrage/src/arbitarger.ts:33]: is it still needed in prod ? or is it jsut a workaround for the tests ?

File: packages/bot-arbitrage/src/arbitarger.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  arbitragerContract: MgvArbitrage,
+  arbParams: ArbParams
+) => {
+  const gasEstimate = await arbitragerContract.estimateGas[arbParams.method](
+    arbParams
+  );
+
+  const tx = await arbitragerContract[arbParams.method](arbParams, {
+    gasLimit: gasEstimate,
+  });
```
Comment (maxencerb) [inline packages/bot-arbitrage/src/arbitarger.ts:127]: Same as above, is it needed still ?

Comment (Louis-Amas) [reply packages/bot-arbitrage/src/arbitarger.ts:127]: Yep we need to estimate the gas

File: packages/bot-arbitrage/src/arbitarger.ts (hunk 1 above)
Comment (Louis-Amas) [reply packages/bot-arbitrage/src/arbitarger.ts:21]: Yes

File: packages/bot-arbitrage/src/arbitarger.ts (hunk 2 above)
Comment (Louis-Amas) [reply packages/bot-arbitrage/src/arbitarger.ts:33]: I needs for both I guess but I reduce it to multiply by two. It should be ok as we called this once per token

File: packages/mangrove-arbitrage/src/MgvArbitrage.sol [hunk 4]
```diff
@@ -1,55 +1,82 @@
...
+  /// copied from v3-core repo to interact with low level swap function of uniwap
+  uint160 internal constant MIN_SQRT_RATIO = 4295128739;
+  uint160 internal constant MAX_SQRT_RATIO = 1461446703485210103287273052203988822378723970342;
+
   IMangrove public mgv;
-  address public arbitrager;
+
+  bool isArbitraging;
+
+  mapping(address => bool) public pools;
```
Comment (maxencerb) [inline packages/mangrove-arbitrage/src/MgvArbitrage.sol]: Sorry but can you natspec these, it's not very clear what the address is referring to at this point.

File: packages/mangrove-arbitrage/src/MgvArbitrage.sol [hunk 5]
```diff
@@ -1,55 +1,82 @@
...
-  /// @param newArbitrager The new arbitrager
-  function setArbitrager(address newArbitrager) external onlyAdmin {
-    arbitrager = newArbitrager;
+    if (token0 != takerGivesToken) {
+      // price is in takerWantsToken
+      amount = 10 ** token0Decimals * balance / price;
+    } else {
+      // price is in takerGivesToken
+      amount = (balance * price) / 10 ** token0Decimals;
+    }
```
Comment (maxencerb) [inline packages/mangrove-arbitrage/src/MgvArbitrage.sol:73]: Assuming infinite liquidity on uniswap. Do we know for what volume is the price true ? Can't we estimate the precise output of a swap instead ?

Comment (Louis-Amas) [reply packages/mangrove-arbitrage/src/MgvArbitrage.sol:73]: No but we don't really care,
This is just to give us a signal if we should either SELL all token A or BUY exact amount of token B. As we are arbitraging liquid pools it should not be an issue. 

The worst case scenario would be that we miss one arbitrage opportunity.

File: packages/bot-utils/package.json [hunk 6]
```diff
@@ -30,7 +30,7 @@
   },
   "dependencies": {
     "@ethersproject/experimental": "^5.7.0",
-    "@mangrovedao/mangrove.js": "1.4.21",
+    "@mangrovedao/mangrove.js": "2.0.0-1",
```
Comment (maxencerb) [inline packages/bot-utils/package.json:33]: We should upgrade to latest version when merging our PRs

**Review by maxencerb — APPROVED** [review]
If all test passes, seems alright to me. Also if all past comments are resolved, all good

**Pull request closed without merging**